src/accumulation.jl

```diff
@@ -38,30 +38,38 @@ end
     is_inplaceable_destination(x) -> Bool
 
 Returns true if `x` is suitable for for storing inplace accumulation of gradients.
-For arrays this boils down `x .= y` if will work to mutate `x`, if `y` is an appropriate
-tangent.
-Wrapper array types do not need to overload this if they overload `Base.parent`, and are
-`is_inplaceable_destination` if and only if their parent array is.
-Other types should overload this, as it defaults to `false`.
+For arrays this means `x .= y` will mutate `x`, if `y` is an appropriate tangent.
+
+Here "appropriate" means that both are real or both are complex,
+and that for structured matrices like `x isa Diagonal`, `y` shares this structure.
+
+Wrapper array types should overload this function if they can be written into.
+Before ChainRulesCore 1.16, it would guess `true` for most wrappers based on `parent`,
+but this is not safe, e.g. it will lead to an error with ReadOnltArrays.jl. 
+
+There must always be a correct non-mutating path, so in uncertain cases,
+this function returns `false`.
 """
 is_inplaceable_destination(::Any) = false
-is_inplaceable_destination(::Array) = true
+
+is_inplaceable_destination(::DenseArray)= true
+is_inplaceable_destination(::DenseArray{<:Integer}) = false
```

Comment:
The intention was to allow for `CuArray`s. But perhaps this is unsafe, as:
```
julia> Base.Experimental.Const([1,2]) isa DenseArray
true

julia> subtypes(DenseArray)
13-element Vector{Any}:
 Array
 Base.CodeUnits
 Base.Experimental.Const
 CUDA.Const
 CUDA.CuDeviceArray
 GPUArrays.AbstractDeviceArray
 GPUArraysCore.AbstractGPUArray
 Metal.Const
 Metal.MtlLargerDeviceArray
 MtlDeviceArray
 Random.UnsafeView
 SharedArrays.SharedArray
 SparseArrays.CHOLMOD.Dense
```

```suggestion
is_inplaceable_destination(::Array)= true
is_inplaceable_destination(:: Array{<:Integer}) = false
```

Reply:
Or perhaps:
```suggestion
is_inplaceable_destination(x::DenseArray)= ismutable(x)
is_inplaceable_destination(x::DenseArray{<:Integer}) = false
```

Reply:
I think to be truly safe we much only accept `Array`
Since anyone can subtype `DenseArray`, and overload `setindex` however they want

Reply:
A pity to miss CuArrays. (Although previous version also marks them non-writable.) But agree that if the goal is always to fail on the safe side, then is has to be Array.